flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStore.java

```diff
@@ -143,6 +145,10 @@ public void recover() throws Exception {
 		int numberOfInitialCheckpoints = initialCheckpoints.size();
 
 		LOG.info("Found {} checkpoints in ZooKeeper.", numberOfInitialCheckpoints);
+		if (haveAllDownloaded(initialCheckpoints)) {
```

Comment:
What happens if all but one checkpoint have been downloaded? Would it download all of them or just the missing one?
If the former, I'd modify this method to return all missing checkpoints instead.

Reply:
Yes, if at least one checkpoint is missing (or some extra checkpoints exist) it will re-download `initialCheckpoints`.
It's quite a sensitive part affecting correctness so I think it's better to keep it as simple as possible and not optimize further.

With that in mind, I think the only use of returning missing checkpoints would be to log them. 
Is this what you propose?

Reply:
I was thinking of optimization, but I understand your argument.